RecoEgamma/ElectronIdentification/src/ElectronMVAEstimatorRun2.cc

```diff
@@ -0,0 +1,196 @@
+               << " not found in variable definition file!" << std::endl;
+        }
+        variables_[i].push_back(index);
+
+    }
+  }
+}
+
+ElectronMVAEstimatorRun2::
+~ElectronMVAEstimatorRun2(){
+}
+
+void ElectronMVAEstimatorRun2::setConsumes(edm::ConsumesCollector&& cc) const {
+  // All tokens for event content needed by this MVA
+  // Tags from the variable helper
+  for (auto &tag : mvaVarMngr_.getHelperInputTags()) {
+      cc.consumes<edm::ValueMap<float>>(tag);
+  }
+  for (auto &tag : mvaVarMngr_.getGlobalInputTags()) {
+      cc.consumes<double>(tag);
+  }
+}
+
+float ElectronMVAEstimatorRun2::
+mvaValue( const edm::Ptr<reco::Candidate>& particle, const edm::Event& iEvent) const {
+  // Try to cast the particle into a reco particle.
+  // This should work for both reco and pat.
+  const edm::Ptr<reco::GsfElectron> eleRecoPtr = ( edm::Ptr<reco::GsfElectron> )particle;
+  if( eleRecoPtr.get() == nullptr ) {
+    throw cms::Exception("MVA failure: ")
+      << " given particle is expected to be reco::GsfElectron or pat::Electron," << std::endl
+      << " but appears to be neither" << std::endl;
+  }
+
+  const int iCategory = findCategory( eleRecoPtr );
+  const std::vector<float> vars = fillMVAVariables( particle, iEvent, iCategory );
+  return mvaValue(iCategory, vars);
+}
+
+float ElectronMVAEstimatorRun2::
+mvaValue( const edm::Ptr<reco::GsfElectron>& particle, const edm::EventBase & iEvent) const {
+
+  const int iCategory = findCategory( particle );
+  const std::vector<float> vars = fillMVAVariables( particle, iEvent, iCategory );
+  return mvaValue(iCategory, vars);
+}
+
+float ElectronMVAEstimatorRun2::
+mvaValue( const int iCategory, const std::vector<float> & vars) const  {
+
+  if(debug_) {
+    std::cout << " *** Inside " << name_ << tag_ << std::endl;
+    std::cout << " category " << iCategory << std::endl;
+    for (int i = 0; i < nVariables_[iCategory]; ++i) {
+        std::cout << " " << mvaVarMngr_.getName(variables_[iCategory][i]) << " " << vars[i] << std::endl;
+    }
+  }
+  const float response = gbrForests_.at(iCategory)->GetResponse(vars.data()); // The BDT score
+
+  if(debug_) {
+    std::cout << " ### MVA " << response << std::endl << std::endl;
+  }
+
+  return response;
+}
+
+int ElectronMVAEstimatorRun2::findCategory( const edm::Ptr<reco::Candidate>& particle) const {
+
+  // Try to cast the particle into a reco particle.
+  // This should work for both reco and pat.
+  const edm::Ptr<reco::GsfElectron> eleRecoPtr = ( edm::Ptr<reco::GsfElectron> )particle;
+  if( eleRecoPtr.get() == nullptr ) {
+    throw cms::Exception("MVA failure: ")
+      << " given particle is expected to be reco::GsfElectron or pat::Electron," << std::endl
+      << " but appears to be neither" << std::endl;
+  }
+  return findCategory(eleRecoPtr);
+}
+
+int ElectronMVAEstimatorRun2::findCategory( const edm::Ptr<reco::GsfElectron>& eleRecoPtr ) const {
+  for (int i = 0; i < nCategories_; ++i) {
```

Comment:
From the looks of the implementation, `findCategory` does not want a `edm::Ptr<reco::GsfElectron>` as an argument, instead it wants a `reco::GsfElectron const&`. The interface should be changed to reflect this.